core/transfer/src/main/java/org/eclipse/dataspaceconnector/transfer/core/transfer/TransferProcessManagerImpl.java

```diff
@@ -423,27 +394,131 @@ private boolean processDeprovisioned(TransferProcess process) {
+        } else if (transferProcess.provisioningComplete()) {
+            transferProcess.transitionProvisioned();
+            transferProcessStore.update(transferProcess);
+            observable.invokeForEach(l -> l.preProvisioned(transferProcess));
+        } else {
+            transferProcessStore.update(transferProcess);
         }
+    }
 
+    private void handleDeprovisionResult(TransferProcess transferProcess, List<DeprovisionResult> results) {
         if (transferProcess.getState() == ERROR.code()) {
-            monitor.severe(format("TransferProcessManager: transfer process %s is in ERROR state, so deprovisioning could not be completed", transferProcess.getId()));
+            monitor.severe(format("TransferProcessManager: transfer process %s is in ERROR state, so deprovisioning could not be processed", transferProcess.getId()));
             return;
         }
 
-        transferProcess.transitionDeprovisioned();
-        transferProcessStore.update(transferProcess);
-        observable.invokeForEach(l -> l.preDeprovisioned(transferProcess));
+        var fatalErrors = new ArrayList<String>();
+        results.forEach(result -> {
+            if (result.failed()) {
+                // Record fatal failure so that transfer process can be transitioned to the error state; if non-fatal, skip
+                var status = result.getFailure().status();
+                if (status == ResponseStatus.FATAL_ERROR) {
+                    fatalErrors.addAll(result.getFailure().getMessages());
+                }
+                return;
+            } else if (result.getContent().isInProcess()) {
+                // Still in process, ignore and continue processing other deprovisioned resources
+                return;
+            }
+            var deprovisionedResource = result.getContent();
+
+            var provisionedResource = transferProcess.getProvisionedResource(deprovisionedResource.getProvisionedResourceId());
+            if (provisionedResource == null) {
+                monitor.severe("Received a deprovision result for a provisioned resource that was not found. Skipping.");
+                return;
+            }
+
+            if (provisionedResource instanceof ProvisionedDataDestinationResource) {
+                // a data destination was provisioned by a consumer
+                var dataDestinationResource = (ProvisionedDataDestinationResource) provisionedResource;
+                if (dataDestinationResource.hasToken()) {
+                    var keyName = dataDestinationResource.getResourceName();
+                    vault.deleteSecret(keyName);
+                }
+            } else if (provisionedResource instanceof ProvisionedContentResource) {
+                var contentResource = (ProvisionedContentResource) provisionedResource;
+                if (contentResource.hasToken()) {
+                    var keyName = contentResource.getResourceName();
+                    vault.deleteSecret(keyName);
```

Comment:
The javadoc of the Vault service indicates that an attempted delete may not be possible. I think we should explicitly handle that case.

Reply:
But here there's no exception checking here, so at this point I would either catch the exception or specifying on the vault interface that the method should not throws an exception (methods that returns `Result` should never throw exceptions)

Reply:
Again, I think we should model this. It seems strange that code would issue a request to delete something that doesn't exist when it can check if it does. We also want to track the fact that a resource is associated with a secret.

Reply:
Looking at this more, I think it still needs to check if a token is associated or not. If it is, it should then check the result and send an error to the monitor if the method returns an error deleting the token. Tracking the token also makes it possible to distinguish this situation from a failure where the token never existed.

I think we should do this in another PR as this one is blocking others I have. I can submit a PR that has this check tomorrow.